// @ennealand/enigraph@0.0.10/dist/source/types.js
export var NodeType;
(function (NodeType) {
    NodeType["Unknown"] = "unknown";
    NodeType["Const"] = "const";
    NodeType["ConstTuple"] = "const.tuple";
    NodeType["ConstStruct"] = "const.struct";
    NodeType["ConstRole"] = "const.role";
    NodeType["ConstNorole"] = "const.norole";
    NodeType["ConstClass"] = "const.class";
    NodeType["ConstAbstract"] = "const.abstract";
    NodeType["ConstMaterial"] = "const.material";
    NodeType["Var"] = "var";
    NodeType["VarTuple"] = "var.tuple";
    NodeType["VarStruct"] = "var.struct";
    NodeType["VarRole"] = "var.role";
    NodeType["VarNorole"] = "var.norole";
    NodeType["VarClass"] = "var.class";
    NodeType["VarAbstract"] = "var.abstract";
    NodeType["VarMaterial"] = "var.material";
})(NodeType || (NodeType = {}));
export var EdgeType;
(function (EdgeType) {
    EdgeType["EdgeConst"] = "EdgeConst";
    EdgeType["ArcConst"] = "ArcConst";
    EdgeType["ArcConstPermPosAccess"] = "AccessConstPosPerm";
    EdgeType["ArcConstPermNegAccess"] = "AccessConstNegPerm";
    EdgeType["ArcConstPermFuzAccess"] = "AccessConstFuzPerm";
})(EdgeType || (EdgeType = {}));
