// @ennealand/enigraph@0.0.10/dist/index.js
import './style.css';import { options as ve, Fragment as q, render as _e } from "preact";
import { useComputed as S, useSignal as $, computed as Xe, effect as Ge, batch as He } from "@preact/signals";
import { useRef as Z, useMemo as O, useCallback as w, useEffect as b } from "preact/hooks";
import { useDeepSignal as ee, deepSignal as Ye } from "deepsignal";
const Ke = "_9vkj", Ie = "e43s", Te = "a3s2", Ve = "x4jx", Be = "nm1e", B = {
  container: Ke,
  highlight: Ie,
  node: Te,
  text: Ve,
  noselect: Be
}, G = (e) => e && typeof e == "object" && "value" in e ? e.value : e, te = (...e) => {
  let o = "";
  for (const n of e)
    if (n) {
      if (typeof n == "string")
        o += o ? " " + n : n;
      else if (typeof n == "object")
        for (const c in n)
          ({}).hasOwnProperty.call(n, c) && n[c] && (o += o ? " " + c : c);
    }
  return o;
};
var Oe = 0;
function t(e, o, n, c, s, u) {
  var l, g, v = {};
  for (g in o)
    g == "ref" ? l = o[g] : v[g] = o[g];
  var x = { type: e, props: v, key: n, ref: l, __k: null, __: null, __b: 0, __e: null, __d: void 0, __c: null, constructor: void 0, __v: --Oe, __i: -1, __u: 0, __source: s, __self: u };
  if (typeof e == "function" && (l = e.defaultProps))
    for (g in l)
      v[g] === void 0 && (v[g] = l[g]);
  return ve.vnode && ve.vnode(x), x;
}
const ye = ({
  type: e,
  noselect: o,
  x: n,
  y: c,
  label: s,
  mousedown: u,
  mouseup: l,
  textDoubleClick: g,
  highlight: v,
  padding: x
}) => t("g", {
  children: [t("g", {
    class: te(B.container, o && B.noselect, v && B.highlight),
    onMouseDown: u,
    onMouseUp: l,
    children: [x && t("circle", {
      cx: n,
      cy: c,
      r: x,
      fill: "transparent"
    }), t("use", {
      class: B.node,
      xlinkHref: `#scg.node.${e}`,
      x: n,
      y: c
    })]
  }), s && t("text", {
    x: n + 17,
    y: c + 21,
    class: B.text,
    onDblClick: g,
    children: s
  })]
});
var E = /* @__PURE__ */ ((e) => (e.Unknown = "unknown", e.Const = "const", e.ConstTuple = "const.tuple", e.ConstStruct = "const.struct", e.ConstRole = "const.role", e.ConstNorole = "const.norole", e.ConstClass = "const.class", e.ConstAbstract = "const.abstract", e.ConstMaterial = "const.material", e.Var = "var", e.VarTuple = "var.tuple", e.VarStruct = "var.struct", e.VarRole = "var.role", e.VarNorole = "var.norole", e.VarClass = "var.class", e.VarAbstract = "var.abstract", e.VarMaterial = "var.material", e))(E || {}), A = /* @__PURE__ */ ((e) => (e.EdgeConst = "EdgeConst", e.ArcConst = "ArcConst", e.ArcConstPermPosAccess = "AccessConstPosPerm", e.ArcConstPermNegAccess = "AccessConstNegPerm", e.ArcConstPermFuzAccess = "AccessConstFuzPerm", e))(A || {});
const Re = "_7bu2", Ue = "_386z", We = "nvzv", Fe = "d84p", Ne = "yjnd", qe = "_4193", m = {
  container: Re,
  highlight: Ue,
  edge: We,
  fill: Fe,
  stroke: Ne,
  noselect: qe
}, Ze = ({
  x1: e,
  y1: o,
  x2: n,
  y2: c
}) => {
  const s = Math.atan2(c - o, n - e);
  return n -= Math.cos(s) * 10, c -= Math.sin(s) * 10, t("g", {
    class: m.edge,
    children: [t("path", {
      "stroke-width": "7.5",
      d: `M ${e} ${o} L ${n} ${c}`,
      class: m.stroke
    }), t("path", {
      "stroke-width": "5",
      class: m.fill,
      d: `M ${e} ${o} L ${n} ${c}`,
      "marker-end": 'url("#end-arrow-common")'
    })]
  });
}, Je = ({
  x1: e,
  y1: o,
  x2: n,
  y2: c
}) => {
  const s = Math.atan2(c - o, n - e);
  return n -= Math.cos(s) * 10, c -= Math.sin(s) * 10, t("g", {
    class: m.edge,
    children: [t("path", {
      class: m.stroke,
      "stroke-width": "1.5",
      d: `M ${e} ${o} L ${n} ${c}`,
      "marker-end": 'url("#end-arrow-access")'
    }), t("path", {
      class: m.stroke,
      "stroke-dasharray": "0, 2, 2, 6",
      "stroke-width": "10px",
      d: `M ${e} ${o} L ${n + 4} ${c - 4}`
    })]
  });
}, Qe = ({
  x1: e,
  y1: o,
  x2: n,
  y2: c
}) => {
  const s = Math.atan2(c - o, n - e);
  return n -= Math.cos(s) * 10, c -= Math.sin(s) * 10, t("g", {
    class: m.edge,
    children: t("path", {
      class: m.stroke,
      "stroke-width": "1.5",
      d: `M ${e} ${o} L ${n} ${c}`,
      "marker-end": 'url("#end-arrow-access")'
    })
  });
}, je = ({
  x1: e,
  y1: o,
  x2: n,
  y2: c
}) => t("g", {
  class: m.edge,
  children: [t("path", {
    "stroke-width": "8",
    d: `M ${e} ${o} L ${n} ${c}`,
    class: m.stroke
  }), t("path", {
    "stroke-width": "5",
    class: m.fill,
    d: `M ${e} ${o} L ${n} ${c}`
  })]
}), pe = ({
  x1: e,
  y1: o,
  x2: n,
  y2: c
}) => t("g", {
  class: m.edge,
  children: [t("path", {
    class: m.stroke,
    "stroke-width": "1.5",
    d: `M ${e} ${o} L ${n} ${c}`,
    "marker-end": 'url("#end-arrow-access")'
  }), t("path", {
    class: m.stroke,
    "stroke-dasharray": "0, 20, 2, 6",
    "stroke-width": "5px",
    d: `M ${e} ${o - 5} L ${n + 5} ${c}`
  }), t("path", {
    class: m.stroke,
    "stroke-dashoffset": "14",
    "stroke-dasharray": "0, 20, 2, 6",
    "stroke-width": "5px",
    d: `M ${e - 5} ${o} L ${n} ${c + 5}`
  })]
}), et = {
  [A.EdgeConst]: je,
  [A.ArcConst]: Ze,
  [A.ArcConstPermPosAccess]: Qe,
  [A.ArcConstPermNegAccess]: Je,
  [A.ArcConstPermFuzAccess]: pe
}, ne = ({
  type: e,
  noselect: o,
  x1: n = 0,
  y1: c = 0,
  x2: s = 0,
  y2: u = 0,
  mousedown: l,
  mouseup: g,
  highlight: v,
  padding: x = 0
}) => {
  const r = et[e], a = Math.sqrt(Math.pow(s - n, 2) + Math.pow(u - c, 2)), d = a && (s - n) / a * x, h = a && (u - c) / a * x;
  return t("g", {
    class: te(m.container, o && m.noselect, v && m.highlight),
    onMouseDown: l,
    onMouseUp: g,
    children: [t("path", {
      d: `M ${n + d} ${c + h} L ${s - d} ${u - h}`,
      "stroke-width": "15",
      stroke: "transparent"
    }), t(r, {
      x1: n + d,
      y1: c + h,
      x2: s - d,
      y2: u - h
    })]
  });
}, tt = () => t("defs", {
  children: [t("circle", {
    id: "scg.node.const.outer",
    cx: "0",
    cy: "0",
    r: "10"
  }), t("rect", {
    id: "scg.node.var.outer",
    x: "-10",
    y: "-10",
    width: "20",
    height: "20"
  }), t("clipPath", {
    id: "scg.node.const.clip",
    children: t("use", {
      xlinkHref: "#scg.node.const.clip"
    })
  }), t("clipPath", {
    id: "scg.node.var.clip",
    children: t("use", {
      xlinkHref: "#scg.node.var.clip"
    })
  }), t("g", {
    id: "scg.node.unknown",
    children: [t("circle", {
      cx: "0",
      cy: "0",
      r: "2.5",
      "stroke-width": "5"
    }), t("text", {
      x: "7",
      y: "15",
      class: "SCgText"
    })]
  }), t("g", {
    id: "scg.node.const",
    children: [t("use", {
      xlinkHref: "#scg.node.const.outer"
    }), t("text", {
      x: "17",
      y: "21",
      class: "SCgText"
    })]
  }), t("g", {
    id: "scg.node.const.tuple",
    children: [t("use", {
      xlinkHref: "#scg.node.const.outer"
    }), t("line", {
      x1: "-10",
      x2: "10",
      y1: "0",
      y2: "0",
      "stroke-width": "1.3"
    }), t("text", {
      x: "17",
      y: "21",
      class: "SCgText"
    })]
  }), t("g", {
    id: "scg.node.const.struct",
    children: [t("use", {
      xlinkHref: "#scg.node.const.outer"
    }), t("circle", {
      cx: "0",
      cy: "0",
      r: "1",
      "stroke-width": "2"
    }), t("text", {
      x: "17",
      y: "21",
      class: "SCgText"
    })]
  }), t("g", {
    id: "scg.node.const.role",
    children: [t("use", {
      xlinkHref: "#scg.node.const.outer"
    }), t("line", {
      x1: "0",
      x2: "0",
      y1: "-10",
      y2: "10",
      "stroke-width": "1.3"
    }), t("line", {
      x1: "-10",
      x2: "10",
      y1: "0",
      y2: "0",
      "stroke-width": "1.3"
    }), t("text", {
      x: "17",
      y: "21",
      class: "SCgText"
    })]
  }), t("g", {
    id: "scg.node.const.norole",
    children: [t("use", {
      xlinkHref: "#scg.node.const.outer"
    }), t("line", {
      x1: "-10",
      x2: "10",
      y1: "0",
      y2: "0",
      "stroke-width": "1.3",
      transform: "rotate(45, 0, 0)"
    }), t("line", {
      x1: "-10",
      x2: "10",
      y1: "0",
      y2: "0",
      "stroke-width": "1.3",
      transform: "rotate(-45, 0, 0)"
    }), t("text", {
      x: "17",
      y: "21",
      class: "SCgText"
    })]
  }), t("g", {
    id: "scg.node.const.class",
    children: [t("use", {
      xlinkHref: "#scg.node.const.outer"
    }), t("line", {
      x1: "-10",
      x2: "10",
      y1: "-3",
      y2: "-3",
      "stroke-width": "1.3"
    }), t("line", {
      x1: "-10",
      x2: "10",
      y1: "3",
      y2: "3",
      "stroke-width": "1.3"
    }), t("line", {
      x1: "-3",
      x2: "-3",
      y1: "-10",
      y2: "10",
      "stroke-width": "1.3"
    }), t("line", {
      x1: "3",
      x2: "3",
      y1: "-10",
      y2: "10",
      "stroke-width": "1.3"
    }), t("text", {
      x: "17",
      y: "21",
      class: "SCgText"
    })]
  }), t("g", {
    id: "scg.node.const.abstract",
    children: [t("use", {
      xlinkHref: "#scg.node.const.outer"
    }), t("g", {
      "stroke-width": "1.3",
      children: [t("line", {
        x1: "-10",
        x2: "10",
        y1: "-6",
        y2: "-6"
      }), t("line", {
        x1: "-10",
        x2: "10",
        y1: "-3",
        y2: "-3"
      }), t("line", {
        x1: "-10",
        x2: "10",
        y1: "0",
        y2: "0"
      }), t("line", {
        x1: "-10",
        x2: "10",
        y1: "3",
        y2: "3"
      }), t("line", {
        x1: "-10",
        x2: "10",
        y1: "6",
        y2: "6"
      })]
    }), t("text", {
      x: "17",
      y: "21",
      class: "SCgText"
    })]
  }), t("g", {
    id: "scg.node.const.material",
    children: [t("use", {
      xlinkHref: "#scg.node.const.outer"
    }), t("g", {
      "stroke-width": "1.3",
      transform: "rotate(-45, 0, 0)",
      children: [t("line", {
        x1: "-9",
        x2: "9",
        y1: "-6",
        y2: "-6"
      }), t("line", {
        x1: "-10",
        x2: "10",
        y1: "-3",
        y2: "-3"
      }), t("line", {
        x1: "-10",
        x2: "10",
        y1: "0",
        y2: "0"
      }), t("line", {
        x1: "-10",
        x2: "10",
        y1: "3",
        y2: "3"
      }), t("line", {
        x1: "-9",
        x2: "9",
        y1: "6",
        y2: "6"
      })]
    }), t("text", {
      x: "17",
      y: "21",
      class: "SCgText"
    })]
  }), t("g", {
    id: "scg.node.var",
    children: [t("use", {
      xlinkHref: "#scg.node.var.outer"
    }), t("text", {
      x: "17",
      y: "21",
      class: "SCgText"
    })]
  }), t("g", {
    id: "scg.node.var.tuple",
    children: [t("use", {
      xlinkHref: "#scg.node.var.outer"
    }), t("line", {
      x1: "-10",
      x2: "10",
      y1: "0",
      y2: "0",
      "stroke-width": "1.3"
    }), t("text", {
      x: "17",
      y: "21",
      class: "SCgText"
    })]
  }), t("g", {
    id: "scg.node.var.struct",
    children: [t("use", {
      xlinkHref: "#scg.node.var.outer"
    }), t("circle", {
      cx: "0",
      cy: "0",
      r: "1",
      "stroke-width": "2"
    }), t("text", {
      x: "17",
      y: "21",
      class: "SCgText"
    })]
  }), t("g", {
    id: "scg.node.var.role",
    children: [t("use", {
      xlinkHref: "#scg.node.var.outer"
    }), t("line", {
      x1: "0",
      x2: "0",
      y1: "-10",
      y2: "10",
      "stroke-width": "1.3"
    }), t("line", {
      x1: "-10",
      x2: "10",
      y1: "0",
      y2: "0",
      "stroke-width": "1.3"
    }), t("text", {
      x: "17",
      y: "21",
      class: "SCgText"
    })]
  }), t("g", {
    id: "scg.node.var.norole",
    children: [t("use", {
      xlinkHref: "#scg.node.var.outer"
    }), t("line", {
      x1: "-12",
      x2: "12",
      y1: "0",
      y2: "0",
      "stroke-width": "1.3",
      transform: "rotate(45, 0, 0)"
    }), t("line", {
      x1: "-12",
      x2: "12",
      y1: "0",
      y2: "0",
      "stroke-width": "1.3",
      transform: "rotate(-45, 0, 0)"
    }), t("text", {
      x: "17",
      y: "21",
      class: "SCgText"
    })]
  }), t("g", {
    id: "scg.node.var.class",
    children: [t("use", {
      xlinkHref: "#scg.node.var.outer"
    }), t("line", {
      x1: "-10",
      x2: "10",
      y1: "-3",
      y2: "-3",
      "stroke-width": "1.3"
    }), t("line", {
      x1: "-10",
      x2: "10",
      y1: "3",
      y2: "3",
      "stroke-width": "1.3"
    }), t("line", {
      x1: "-3",
      x2: "-3",
      y1: "-10",
      y2: "10",
      "stroke-width": "1.3"
    }), t("line", {
      x1: "3",
      x2: "3",
      y1: "-10",
      y2: "10",
      "stroke-width": "1.3"
    }), t("text", {
      x: "17",
      y: "21",
      class: "SCgText"
    })]
  }), t("g", {
    id: "scg.node.var.abstract",
    children: [t("use", {
      xlinkHref: "#scg.node.var.outer"
    }), t("g", {
      "stroke-width": "1.3",
      children: [t("line", {
        x1: "-10",
        x2: "10",
        y1: "-6",
        y2: "-6"
      }), t("line", {
        x1: "-10",
        x2: "10",
        y1: "-3",
        y2: "-3"
      }), t("line", {
        x1: "-10",
        x2: "10",
        y1: "0",
        y2: "0"
      }), t("line", {
        x1: "-10",
        x2: "10",
        y1: "3",
        y2: "3"
      }), t("line", {
        x1: "-10",
        x2: "10",
        y1: "6",
        y2: "6"
      })]
    }), t("text", {
      x: "17",
      y: "21",
      class: "SCgText"
    })]
  }), t("g", {
    id: "scg.node.var.material",
    children: [t("use", {
      xlinkHref: "#scg.node.var.outer"
    }), t("g", {
      "stroke-width": "1.3",
      transform: "rotate(-45, 0, 0)",
      children: [t("line", {
        x1: "-9",
        x2: "9",
        y1: "-6",
        y2: "-6"
      }), t("line", {
        x1: "-11",
        x2: "11",
        y1: "-3",
        y2: "-3"
      }), t("line", {
        x1: "-13",
        x2: "13",
        y1: "0",
        y2: "0"
      }), t("line", {
        x1: "-11",
        x2: "11",
        y1: "3",
        y2: "3"
      }), t("line", {
        x1: "-9",
        x2: "9",
        y1: "6",
        y2: "6"
      })]
    }), t("text", {
      x: "17",
      y: "21",
      class: "SCgText"
    })]
  }), t("marker", {
    id: "end-arrow-access",
    viewBox: "0 -5 10 10",
    refX: "0",
    markerWidth: "8",
    markerHeight: "14",
    orient: "auto",
    markerUnits: "userSpaceOnUse",
    children: t("path", {
      d: "M0,-4L10,0L0,4",
      fill: "var(--scgEdgeMarkerColor, var(--scgEdgeColor, #000))"
    })
  }), t("marker", {
    id: "end-arrow-common",
    viewBox: "0 -5 10 10",
    refX: "0",
    markerWidth: "10",
    markerHeight: "16",
    orient: "auto",
    markerUnits: "userSpaceOnUse",
    children: t("path", {
      d: "M0,-4L10,0L0,4",
      fill: "var(--scgEdgeMarkerColor, var(--scgEdgeColor, #000))"
    })
  })]
}), nt = "_9ga9", ot = "sy6u", xe = {
  graph: nt,
  innerHtml: ot
}, st = (e, o) => {
  const n = Z(null), c = e && O(() => e / 2, [e]), s = o && O(() => o / 2, [o]), u = w((g) => t(ct, {
    width: e,
    height: o,
    centerX: c,
    centerY: s,
    ...g,
    pref: n
  }), [e, o]), l = w((g, v) => {
    if (!n.current)
      return [0, 0];
    const x = n.current.getBoundingClientRect();
    return [g - x.x - c, v - x.y - s];
  }, [n.current]);
  return {
    BaseGraph: u,
    getInnerPoint: l
  };
}, ct = (e) => {
  const o = G(e.transform), n = G(e.noselect), c = G(e.nolabels);
  return t("div", {
    class: xe.graph,
    "data-movable": G(e.movable) ? "" : void 0,
    "data-moving": o?.moving ? "" : void 0,
    "data-dragging": G(e.dragging) ? "" : void 0,
    "data-selecting": G(e.selecting) ? "" : void 0,
    children: [t("svg", {
      ref: e.pref,
      xmlns: "http://www.w3.org/2000/svg",
      xmlnsXlink: "http://www.w3.org/1999/xlink",
      viewBox: `-${e.centerX} -${e.centerY} ${e.width} ${e.height}`,
      width: `${e.width}px`,
      height: `${e.height}px`,
      onContextMenu: (s) => (s.preventDefault(), s.stopPropagation()),
      onMouseDown: e.onMouseDown,
      onWheel: e.onWheel,
      onMouseEnter: e.onMouseEnter,
      onMouseLeave: e.onMouseLeave,
      children: [t(tt, {}), e.elements && t("g", {
        transform: o && `translate(${o.x} ${o.y}) scale(${o.zoom})`,
        children: [e.before, e.elements.edges.map((s, u) => t(ne, {
          type: s.type,
          x1: s.source.x,
          y1: s.source.y,
          x2: s.target.x,
          y2: s.target.y,
          mousedown: (l) => e.onEdgeMouseDown?.(l, s, u),
          mouseup: (l) => e.onEdgeMouseUp?.(l, s, u),
          highlight: G(e.highlight)?.has(s.id),
          noselect: n && (n === !0 || n.has(s.id)),
          padding: e.padding
        }, s.id)), e.elements.nodes.map((s, u) => t(ye, {
          type: s.type,
          x: Math.round(s.x) || 0,
          y: Math.round(s.y) || 0,
          label: c && (c === !0 || c.has(s.id)) ? void 0 : s.label,
          mousedown: (l) => e.onNodeMouseDown?.(l, s, u),
          mouseup: (l) => e.onNodeMouseUp?.(l, s, u),
          highlight: G(e.highlight)?.has(s.id),
          noselect: n && (n === !0 || n.has(s.id)),
          padding: e.padding && e.padding + 1,
          textDoubleClick: () => e.onTextDoubleClick?.(s)
        }, s.id)), e.inner]
      }), e.children]
    }), t("div", {
      class: xe.innerHtml,
      style: {
        transform: o && `translate(${o.x}px, ${o.y}px) scale(${o.zoom}) translate(50%, 50%)`
      },
      onWheel: e.onWheel,
      children: e.innerHtml
    })]
  });
}, at = (e) => {
  const o = ee({
    values: []
  }), n = (v, x, r) => {
    const a = {
      id: 0,
      type: r,
      x: v,
      y: x
    };
    if (e.addNode(a), o.values.length) {
      for (const {
        type: d,
        source: h
      } of o.values)
        e.addEdge({
          id: 0,
          type: d,
          source: h,
          target: a
        });
      o.values = [];
    }
  }, c = (v, x, r) => {
    if (!e.selection)
      return;
    const a = [];
    for (const d of e.nodes)
      e.selection.value.has(d.id) && a.push({
        type: r,
        source: d,
        target: {
          x: v,
          y: x
        }
      });
    o.values = a;
  }, s = (v) => {
    if (!e.selection || !o.values.length)
      return;
    const [x, r] = e.localize(...e.getInnerPoint(v.clientX, v.clientY));
    for (const a of o.values)
      a.target.x = x, a.target.y = r;
  }, u = (v, x) => {
    if (o.values.length) {
      for (const {
        type: r,
        source: a
      } of o.values)
        for (const d of x)
          v.has(d.id) && e.addEdge({
            id: 0,
            type: r,
            source: a,
            target: d
          });
      o.values = [];
    }
  }, l = w(() => e.Edge ? t(rt, {
    Edge: e.Edge,
    edges: o.values
  }) : null, [o]), g = S(() => !!o.$values?.value.length);
  return {
    createNode: n,
    startDrawingEdge: c,
    updateDrawingEdges: s,
    createEdges: u,
    DrawingEdges: l,
    isDrawingEdges: g
  };
}, rt = (e) => t("g", {
  children: e.edges.map((o, n) => t(e.Edge, {
    type: o.type,
    x1: o.source.x,
    y1: o.source.y,
    x2: o.target.x,
    y2: o.target.y
  }, n))
}), lt = "h6sx", it = "fu9u", ut = {
  disk: lt,
  show: it
}, dt = (e) => t("g", {
  transform: `translate(${e.x} ${e.y})`,
  children: t("g", {
    class: ut.disk,
    children: e.type === "node" ? e.nodeOptions.map(({
      type: o,
      x1: n,
      y1: c,
      x2: s,
      y2: u,
      textX: l,
      textY: g,
      nodeX: v,
      nodeY: x
    }, r) => t("g", {
      onMouseDown: (a) => a.stopPropagation(),
      onMouseUp: (a) => e.click(e.type, e.x, e.y, a, o),
      children: [t("path", {
        d: `M ${n} ${c} A 70 70 0 0 1 ${s} ${u}`,
        "stroke-width": "90"
      }), t("text", {
        x: l,
        y: g,
        "stroke-width": "90",
        children: r + 1
      }), t(ye, {
        x: v,
        y: x,
        type: o,
        noselect: !0
      })]
    }, o)) : e.edgeOptions.map(({
      type: o,
      x1: n,
      y1: c,
      x2: s,
      y2: u,
      textX: l,
      textY: g,
      edgeX1: v,
      edgeY1: x,
      edgeX2: r,
      edgeY2: a
    }, d) => t("g", {
      onMouseDown: (h) => h.stopPropagation(),
      onMouseUp: (h) => e.click(e.type, e.x, e.y, h, o),
      children: [t("path", {
        d: `M ${n} ${c} A 70 70 0 0 1 ${s} ${u}`,
        "stroke-width": "90"
      }), t("text", {
        x: l,
        y: g,
        "stroke-width": "90",
        children: d + 1
      }), t(ne, {
        x1: v,
        y1: x,
        x2: r,
        y2: a,
        type: o,
        noselect: !0
      })]
    }, o))
  })
}), ht = [E.Const, E.ConstClass, E.ConstStruct, E.ConstTuple, E.Unknown, E.VarTuple, E.VarStruct, E.VarClass, E.Var], gt = [A.ArcConstPermNegAccess, A.ArcConstPermPosAccess, A.ArcConst, A.EdgeConst], vt = (e) => e.map((o, n, c) => {
  const s = 2 * Math.PI / c.length;
  return {
    type: o,
    x1: Math.round(700 * Math.sin(s * n)) / 10,
    y1: Math.round(700 * -Math.cos(s * n)) / 10,
    x2: Math.round(700 * Math.sin(s * (n + 1))) / 10,
    y2: Math.round(700 * -Math.cos(s * (n + 1))) / 10,
    textX: Math.round(990 * Math.sin(0.12 + s * n)) / 10 - 4,
    textY: Math.round(990 * -Math.cos(0.12 + s * n)) / 10 + 6,
    nodeX: Math.round(740 * Math.sin(0.35 + s * n)) / 10,
    nodeY: Math.round(740 * -Math.cos(0.35 + s * n)) / 10
  };
}), xt = (e) => e.map((o, n, c) => {
  const s = 2 * Math.PI / c.length;
  return {
    type: o,
    x1: Math.round(700 * Math.sin(s * n)) / 10,
    y1: Math.round(700 * -Math.cos(s * n)) / 10,
    x2: Math.round(700 * Math.sin(s * (n + 1))) / 10,
    y2: Math.round(700 * -Math.cos(s * (n + 1))) / 10,
    textX: Math.round(990 * Math.sin(0.12 + s * n)) / 10 - 4,
    textY: Math.round(990 * -Math.cos(0.12 + s * n)) / 10 + 6,
    edgeX1: Math.round(450 * Math.sin(s * (n + 0.5))) / 10,
    edgeY1: Math.round(450 * -Math.cos(s * (n + 0.5))) / 10,
    edgeX2: Math.round(900 * Math.sin(s * (n + 0.5))) / 10,
    edgeY2: Math.round(900 * -Math.cos(s * (n + 0.5))) / 10
  };
}), ft = (e, o) => {
  const n = O(() => vt(o?.nodeTypes ?? ht), [o?.nodeTypes]), c = O(() => xt(o?.edgeTypes ?? gt), [o?.edgeTypes]), s = ee({
    x: 0,
    y: 0,
    type: "node",
    shown: !1
  }), u = w((a) => {
    if (o?.getInnerPoint) {
      const [d, h] = o.getInnerPoint(a.clientX, a.clientY);
      if (Math.sqrt((s.x - d) ** 2 + (s.y - h) ** 2) <= 5)
        return;
    }
    s.shown = !1, g();
  }, []), l = w((a) => {
    a.code === "Escape" && (s.shown = !1, g());
  }, []), g = () => {
    document.removeEventListener("keydown", l), document.removeEventListener("mouseup", u);
  };
  return b(() => g, []), {
    Disk: () => s.shown ? t(dt, {
      x: s.x,
      y: s.y,
      type: s.type,
      click: e,
      nodeOptions: n,
      edgeOptions: c
    }) : null,
    showDisk: (a, d, h) => {
      s.type = a, s.x = d, s.y = h, s.shown = !0, document.addEventListener("mouseup", u, {
        once: !0
      }), document.addEventListener("keydown", l, {
        once: !0
      });
    },
    hideDisk: () => {
      s.shown = !1;
    },
    isDiskOpened: s.$shown
  };
}, yt = (e) => {
  const o = $(!1), n = $({
    x: 0,
    y: 0
  }), c = $({
    x: 0,
    y: 0
  }), s = (v) => {
    o.value = !0;
    const [x, r] = e.getInnerPoint(v.clientX, v.clientY);
    n.value.x = x, n.value.y = r, c.value.x = 0, c.value.y = 0, document.addEventListener("mouseup", g, {
      once: !0
    });
  }, u = (v) => {
    const [x, r] = e.getInnerPoint(v.clientX, v.clientY), a = n.value.x - x, d = n.value.y - r, h = e.zoom?.value ?? 1;
    for (const f of e.nodes)
      e.selection.value.has(f.id) && e.changeNodePosition?.(f, f.x - a / h, f.y - d / h);
    c.value.x += a, c.value.y += d, n.value.x = x, n.value.y = r;
  }, l = (v) => {
    if (o.value) {
      if (v?.revert) {
        const x = e.zoom?.value ?? 1;
        for (const r of e.nodes)
          e.selection.value.has(r.id) && e.changeNodePosition?.(r, r.x + c.value.x / x, r.y + c.value.y / x);
      }
      g();
    }
  }, g = w(() => {
    if (document.removeEventListener("mouseup", g), e.nodePositionChanged)
      for (const v of e.nodes)
        e.selection.value.has(v.id) && e.nodePositionChanged(v);
    o.value = !1;
  }, [o, e.nodes, e.selection, e.nodePositionChanged]);
  return b(() => () => document.removeEventListener("mouseup", g), [o, e.nodes, e.selection, e.nodePositionChanged]), {
    isDragging: Xe(() => o.value),
    startDragginig: s,
    updateDragging: u,
    abortDragging: l
  };
}, me = (e, o) => {
  let n = 1 / 0, c = -1 / 0, s = 1 / 0, u = -1 / 0;
  for (const l of e)
    o.has(l.id) && (l.x > u && (u = l.x), l.x < s && (s = l.x), l.y < n && (n = l.y), l.y > c && (c = l.y));
  return {
    top: n,
    bottom: c,
    right: u,
    left: s
  };
}, mt = "_9uvf", wt = "kcyn", kt = "_2acs", Mt = "fyh8", $t = "r5g4", Ct = "yp31", K = {
  placeholder: mt,
  group: wt,
  opened: kt,
  selected: Mt,
  circle: $t,
  indicator: Ct
}, Pt = (e) => O(() => t("g", {
  onMouseDown: (n) => e.onMouseDown?.(n, e.id),
  children: [t("rect", {
    class: te(K.group, e.opened && K.opened, e.selected && K.selected),
    x: e.left - 30,
    y: e.top - 30,
    width: e.right - e.left + 30 * 2,
    height: e.bottom - e.top + 30 * 2
  }), e.indicator && t(q, {
    children: [t("circle", {
      class: K.circle,
      cx: e.left - 25,
      cy: e.top - 25,
      r: 15
    }), t("text", {
      class: K.indicator,
      x: e.left - 29.8,
      y: e.top - 19.2,
      children: e.indicator
    })]
  })]
}), Object.values(e)), Dt = (e) => {
  const o = $(/* @__PURE__ */ new Set()), n = $(null);
  b(() => Ge(() => {
    for (const r of e.groups) {
      const a = me(e.nodes, r.values);
      for (const [d, h] of Object.entries(a))
        r.position[d] !== h && (r.position[d] = h);
    }
  }), [e.nodes, e.groups]);
  const c = (r) => {
    o.value.add(r), o.value = new Set(o.value), r === n.value && (n.value = null);
  }, s = (r) => {
    o.value.delete(r), o.value = new Set(o.value);
  }, u = () => {
    o.value = /* @__PURE__ */ new Set();
  }, l = (r) => n.value = r, g = () => n.value = null, v = S(() => n.value && e.groups.find(({
    id: r
  }) => r === n.value)?.values);
  return {
    Group: w((r) => t("g", {
      class: r.placeholder && K.placeholder,
      children: e.groups.map((a) => t(Pt, {
        id: a.id,
        ...a.position,
        onMouseDown: r.onMouseDown,
        opened: !r.nohighlight && o.value?.has(a.id),
        selected: r.customSelection ? r.customSelection.has(a.id) : !r.nohighlight && n.value === a.id,
        indicator: r.customIndicators?.get(a.id)
      }, a.id))
    }), [e.groups]),
    openGroup: c,
    closeGroup: s,
    closeAllGroups: u,
    selectGroup: l,
    deselectGroup: g,
    selectedGroup: v,
    selectedGroupId: n
  };
}, Et = "am5j", St = "rudp", At = "w16c", p = {
  menu: Et,
  hidden: St,
  button: At
}, zt = (e) => e.buttons.length ? t("div", {
  class: `${p.menu} ${e.shown ? "" : p.hidden}`,
  style: {
    transform: `translate(${e.x}px, ${e.y}px) translate(-50%, 18px)`
  },
  onMouseLeave: e.leave,
  children: e.buttons.map((o, n) => t("div", {
    class: p.button,
    onClick: (c) => o.action(c, e.x, e.y),
    children: o.content
  }, n))
}) : null, Lt = (e) => {
  const o = $(!1), n = Z(0), c = S(() => {
    if (!e.selection.value.size || !e.visible.value)
      return null;
    const l = {
      x: -1 / 0,
      y: -1 / 0
    };
    for (const g of e.nodes)
      e.selection.value.has(g.id) && (g.y < l.y || (l.x = g.x, l.y = g.y));
    return l;
  });
  b(() => {
    e.selection.value?.size && e.visible.value && (o.value = !1, self.clearTimeout(n.current), n.current = self.setTimeout(() => {
      o.value = !0, document.addEventListener("mousedown", s, {
        once: !0,
        capture: !0
      });
    }, 600));
  }, [c.value]);
  const s = w(() => {
    o.value = !1;
  }, []);
  return b(() => document.removeEventListener("mousedown", s, {
    capture: !0
  }), []), {
    Menu: () => c.value ? t(zt, {
      ...c.value,
      buttons: e.buttons.value,
      leave: s,
      shown: o.value,
      padding: e.padding ?? 0
    }) : null,
    hideMenu: s,
    menuNodePosition: c
  };
}, bt = (e) => {
  const o = e.width / 2, n = e.height / 2, c = ee({
    x: 0,
    y: 0,
    zoom: 1,
    moving: !1
  }), s = (a, d) => [(a - c.x) / c.zoom, (d - c.y) / c.zoom], u = (a, d) => [a * c.zoom + c.x, d * c.zoom + c.y], l = $(null), g = (a) => {
    const [d, h] = e.getInnerPoint(a.clientX, a.clientY);
    l.value = {
      x: d,
      y: h
    };
  }, v = (a) => {
    const [d, h] = e.getInnerPoint(a.clientX, a.clientY);
    l.value = {
      x: d,
      y: h
    };
  }, x = () => {
    l.value = null;
  };
  return {
    transform: c,
    centerX: o,
    centerY: n,
    localize: s,
    globalize: u,
    onwheel: (a) => {
      if (a.preventDefault(), a.ctrlKey) {
        let d = a.deltaY * 0.01;
        c.zoom - d < 0.1 ? d = c.zoom - 0.1 : c.zoom - d > 5 && (d = c.zoom - 5), c.x += d / c.zoom * (a.offsetX - o - c.x), c.y += d / c.zoom * (a.offsetY - n - c.y), c.zoom -= d;
      } else
        c.x -= a.deltaX, c.y -= a.deltaY;
    },
    zoom: c.$zoom,
    startMoving: g,
    updateMoving: v,
    stopMoving: x
  };
}, _t = "xm7w", Xt = "x6ar", fe = {
  component: _t,
  input: Xt
}, Gt = (e) => {
  const o = $(null), n = (u) => {
    o.value = {
      node: u
    };
  }, c = () => {
    o.value = null;
  };
  return {
    RenamingArea: w(() => o.value && t(Ht, {
      x: Math.round(o.value.node.x) || 0,
      y: Math.round(o.value.node.y) || 0,
      value: o.value.node.label ?? "",
      submit: (u) => o.value && (e.submit?.(o.value.node, u), c())
    }), [o]),
    startRenaming: n,
    stopRenaming: c,
    isRenaming: o
  };
}, Ht = (e) => {
  const o = Z(null);
  b(() => o.current?.focus(), []);
  const n = (c) => {
    c.key === "Enter" && e.submit(c.currentTarget.value);
  };
  return t("div", {
    class: fe.component,
    style: {
      transform: `translate(${e.x + 19 - 5.8 - 3.5}px, ${e.y - 11 - 5.8}px)`
    },
    children: t("input", {
      ref: o,
      class: fe.input,
      type: "text",
      value: e.value,
      onKeyDown: n,
      onBlur: (c) => e.submit(c.currentTarget.value)
    })
  });
}, Yt = (e) => t("rect", {
  x: Math.min(e.x1, e.x2),
  y: Math.min(e.y1, e.y2),
  width: Math.abs(e.x1 - e.x2),
  height: Math.abs(e.y1 - e.y2),
  rx: "1",
  ry: "1",
  "stroke-width": "1",
  fill: "#0048b61a",
  stroke: "#2669cf",
  "pointer-events": "none"
}), Kt = (e) => {
  const o = e.localize ?? ((r, a) => [r, a]), n = $(null), c = $(/* @__PURE__ */ new Set()), s = $(/* @__PURE__ */ new Set()), u = $(null), l = w(() => {
    s.value.size && (s.value = /* @__PURE__ */ new Set());
  }, [s]), g = (r, a) => {
    const [d, h] = e.getInnerPoint(r.clientX, r.clientY), [f, Y] = o(d, h) ?? [d, h], C = e.padding ?? 15, D = e.nodes.findLast(({
      x: z,
      y: L
    }) => Math.sqrt((z - f) ** 2 + (L - Y) ** 2) <= C), H = a?.clear ? /* @__PURE__ */ new Set() : new Set(s.value);
    if (D) {
      const z = (L) => {
        !a?.deselection && !(e.inversion && !a?.selection && L.has(D.id)) ? L.add(D.id) : L.delete(D.id);
      };
      a?.clear ? u.value = D.id : z(H);
    }
    const _ = u.value && s.value.has(u.value);
    He(() => {
      n.value = Ye({
        x1: d,
        y1: h,
        x2: d,
        y2: h
      }), c.value = _ ? new Set(s.value) : H, a?.clear && !_ && (s.value = /* @__PURE__ */ new Set());
    }), document.addEventListener("mouseup", x, {
      once: !0
    });
  }, v = (r, a) => {
    if (!n.value)
      return;
    if (u.value) {
      s.value.add(u.value), u.value = null, n.value = null;
      return;
    }
    const [d, h] = e.getInnerPoint(r.clientX, r.clientY);
    n.value.x2 = d, n.value.y2 = h;
    const f = e.padding ?? 16, [Y, C] = o(n.value.x1, n.value.y1), [D, H] = o(n.value.x2, n.value.y2), _ = Math.min(Y, D) - f, z = Math.max(Y, D) + f, L = Math.min(C, H) - f, J = Math.max(C, H) + f, k = /* @__PURE__ */ new Set();
    for (const R of e.nodes.keys()) {
      const P = e.nodes.at(-R - 1);
      P.x >= _ && P.x <= z && P.y >= L && P.y <= J ? !a?.deselection && !(e.inversion && !a?.selection && s.value.has(P.id)) && k.add(P.id) : s.value.has(P.id) && k.add(P.id);
    }
    c.value = k;
  }, x = w(() => {
    if (document.removeEventListener("mouseup", x), !n.value)
      return;
    n.value = null;
    let r;
    u.value ? (r = /* @__PURE__ */ new Set(), r.add(u.value), u.value = null) : r = new Set(c.value), e.onSelectionStop?.(r), s.value = r, c.value = /* @__PURE__ */ new Set();
  }, [e.onSelectionStop]);
  return b(() => () => document.removeEventListener("mouseup", x), [x]), {
    /** Area selection rectangle */
    AreaSelection: w(() => n.value && t(Yt, {
      ...n.value
    }), [n]),
    selection: S(() => n.value ? c.value : s.value),
    isSelecting: S(() => !!n.value && (n.value.x1 !== n.value.x2 || n.value.y1 !== n.value.y2)),
    clearSelection: l,
    startSelection: g,
    updateSelection: v
  };
}, It = ({
  elements: e,
  width: o,
  height: n,
  padding: c,
  addNode: s,
  addEdge: u,
  addGroup: l,
  changeNodeLabel: g,
  changeNodePosition: v,
  nodePositionChanged: x,
  removeNode: r,
  edgeTypes: a,
  nodeTypes: d,
  objectSelection: h,
  buttonIcons: f
}) => {
  const {
    BaseGraph: Y,
    getInnerPoint: C
  } = st(o, n), {
    transform: D,
    onwheel: H,
    localize: _,
    globalize: z,
    zoom: L
  } = bt({
    width: o,
    height: n,
    getInnerPoint: C
  }), {
    AreaSelection: J,
    selection: k,
    startSelection: R,
    updateSelection: P,
    clearSelection: we,
    isSelecting: oe
  } = Kt({
    nodes: e.nodes,
    getInnerPoint: C,
    localize: _,
    inversion: !0,
    padding: c,
    onSelectionStop: w((i) => Ae(i, e.nodes), [e.nodes])
  }), {
    Group: se,
    openGroup: ke,
    closeAllGroups: ce,
    selectGroup: Me,
    deselectGroup: $e,
    selectedGroup: Ce,
    selectedGroupId: Pe
  } = Dt({
    nodes: e.nodes,
    groups: e.groups,
    selection: k
  }), {
    RenamingArea: De,
    startRenaming: ae,
    isRenaming: Q
  } = Gt({
    submit: g
  }), re = S(() => Q.value ? /* @__PURE__ */ new Set([Q.value.node.id]) : void 0), le = S(() => h?.values || re.value || Ce.value || k.value), {
    startDragginig: ie,
    updateDragging: ue,
    isDragging: U
  } = yt({
    nodes: e.nodes,
    selection: le,
    getInnerPoint: C,
    zoom: L,
    changeNodePosition: v,
    nodePositionChanged: x
  }), {
    createNode: Ee,
    startDrawingEdge: Se,
    updateDrawingEdges: de,
    createEdges: Ae,
    DrawingEdges: he,
    isDrawingEdges: I
  } = at({
    addNode: s,
    addEdge: u,
    getInnerPoint: C,
    localize: _,
    nodes: e.nodes,
    Edge: w((i) => t(ne, {
      ...i,
      noselect: !0
    }), []),
    selection: k
  }), {
    Disk: ze,
    showDisk: W,
    hideDisk: F,
    isDiskOpened: T
  } = ft((i, y, M, X, j) => {
    if (F(), V.current && V.current.x === y && V.current.y === M) {
      const be = i === "node" ? e.nodes : e.edges;
      for (const ge of be)
        k.value.has(ge.id) && (ge.type = j);
    } else
      i === "node" ? Ee(..._(y, M), j) : Se(y, M, j);
  }, {
    nodeTypes: d,
    edgeTypes: a,
    getInnerPoint: C
  }), V = Z(null), {
    Menu: Le
  } = Lt({
    nodes: e.nodes,
    selection: k,
    visible: S(() => !oe.value && !U.value && !T.value && !I.value && !Q.value),
    buttons: S(() => {
      const i = [{
        content: f ? f.type : t("span", {
          children: "T"
        }),
        action: (y, M, X) => {
          W("node", ...z(M, X)), V.current = {
            x: M,
            y: X
          }, document.addEventListener("mouseup", () => V.current = null, {
            once: !0
          });
        }
      }, {
        content: f ? f.arrow : t("span", {
          children: "A"
        }),
        action: (y, M, X) => W("edge", ...z(M, X))
      }, {
        content: f ? f.group : t("span", {
          children: "G"
        }),
        action: () => l({
          id: 0,
          label: "",
          values: k.value,
          position: me(e.nodes, k.value)
        })
      }];
      if (g && k.value.size === 1) {
        const y = e.nodes.find((M) => k.value.has(M.id));
        y && i.push({
          content: f ? f.rename : t("span", {
            children: "I"
          }),
          action: () => ae(y)
        });
      }
      return r && i.push({
        content: f ? f.delete : t("span", {
          children: "D"
        }),
        action: () => {
          for (const y of k.value)
            r(y);
        }
      }), i.filter((y) => y.content);
    })
  }), N = w((i) => {
    I.value && !T.value && de(i), U.value && ue(i), P(i, {
      deselection: i.altKey,
      selection: i.ctrlKey || i.metaKey
    });
  }, [P, ue, U, I]);
  return b(() => (document.addEventListener("mousemove", N, !0), () => document.removeEventListener("mousemove", N, !0)), [N]), b(() => () => document.removeEventListener("mousemove", N, !0), []), t(Y, {
    width: o,
    height: n,
    centerX: o / 2,
    centerY: n / 2,
    elements: e,
    padding: c,
    transform: D,
    highlight: h ? void 0 : le,
    noselect: !!h || oe,
    dragging: S(() => U.value || I.value),
    onTextDoubleClick: (i) => ae(i),
    nolabels: re,
    onWheel: (i) => {
      H(i), I.value && de(i);
    },
    onMouseDown: (i) => {
      if (!h) {
        if (T && F(), $e(), i.target === i.currentTarget && ce(), i.buttons === 1) {
          R(i, {
            deselection: i.altKey,
            selection: i.ctrlKey || i.metaKey,
            clear: !i.shiftKey && !i.altKey && !i.ctrlKey && !i.metaKey
          });
          return;
        }
        if (i.buttons === 2) {
          if (i.shiftKey)
            return;
          W("node", ...C(i.clientX, i.clientY));
          return;
        }
      }
    },
    onNodeMouseDown: (i, y) => {
      if (!h) {
        if (T && F(), i.buttons === 1) {
          if (i.shiftKey)
            return;
          if (i.ctrlKey || i.metaKey || i.altKey) {
            i.stopPropagation();
            return;
          }
          ie(i);
          return;
        }
        if (i.buttons === 2) {
          if (i.shiftKey)
            return;
          i.stopPropagation(), R(i, {
            clear: !0
          }), W("edge", ...z(y.x, y.y));
        }
      }
    },
    before: t(q, {
      children: t(se, {
        nohighlight: !!h,
        customSelection: h && new Set(h.values),
        onMouseDown: w((i, y) => {
        }, [])
      })
    }),
    inner: t(q, {
      children: [t(se, {
        placeholder: !0,
        nohighlight: !!h,
        customSelection: h?.values,
        customIndicators: h?.indicators,
        onMouseDown: w((i, y) => {
          if (h && h.type === "group") {
            h.action(y);
            return;
          }
          if (T && F(), i.stopPropagation(), ie(i), Pe.value === y) {
            const M = () => {
              ce(), ke(y), document.removeEventListener("mousemove", X);
            }, X = () => {
              document.removeEventListener("mouseup", M);
            };
            document.addEventListener("mouseup", M, {
              once: !0
            }), document.addEventListener("mousemove", X, {
              once: !0
            });
          } else
            Me(y), we();
        }, [h])
      }), he && t(he, {})]
    }),
    innerHtml: t(q, {
      children: [t(Le, {}), t(De, {})]
    }),
    children: [t(ze, {}), t(J, {})]
  });
}, Rt = (e, o) => _e(t(It, {
  ...o
}), document.getElementById(e));
export {
  A as EdgeType,
  It as Graph,
  E as NodeType,
  Rt as default
};
